docs/madr/decisions/006-mesh-traffic-permission.md

```diff
@@ -0,0 +1,503 @@
+# MeshTrafficPermission
+
+* Status: accepted
+
+Technical Story: https://github.com/kumahq/kuma/issues/4222
+
+## Context and Problem Statement
+
+[New policy matching MADR](./005-policy-matching.md) introduces a new approach how Kuma applies configuration to proxies.
+Rolling out strategy implies creating a new set of policies that satisfy a new policy matching.
+Current MADR aims to define a MeshTrafficPermission.
+
+## Considered Options
+
+* Create a MeshTrafficPermission policy
+
+## Decision Outcome
+
+Chosen option: create a MeshTrafficPermission policy
+
+### Overview
+
+MeshTrafficPermission allows blocking unwanted incoming traffic for the service.
+There are might be several reasons why it could be useful to block unwanted traffic:
+* following the [principle of least privilege](https://en.wikipedia.org/wiki/Principle_of_least_privilege)
+* cut off access for clients that shouldn't use your service (legacy clients, clients from another zone, etc.)
+
+### Specification
+
+Top-level targetRef can have all available kinds:
+
+```yaml
+targetRef:
+  kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+  name: ...
+```
+
+MeshTrafficPermission is an inbound policy that's why it has a "from" array:
+
+```yaml
+from:
+  - targetRef:
+      kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+      name: ...
+```
+
+Configuration for MeshTrafficPermission has a single field "action" with 2 values "ALLOW" or "DENY".
+So the overall specification looks like this:
+
+```yaml
+type: MeshTrafficPermission
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh|MeshSubset|MeshService|MeshServiceSubset|MeshGatewayRoute|MeshHTTPRoute
+    name: ...
+  from:
+    - targetRef:
+        kind: Mesh|MeshSubset|MeshService|MeshServiceSubset
+        name: ...
+      conf:
+        action: ALLOW|DENY
```

Comment:
Should there be an option `shadow: true` to run this RBAC rule in shadow (get stats but no enforcement)?

Reply:
Added "shadow_action" field

Reply:
would it be possible to have
```
action: ALLOW|DENY|SHADOW_ALLOW|SHADOW_DENY
```
?

Reply:
Is there any reason to set `action: ALLOW` + `shadow: ALLOW` or `action: DENY` + `shadow: DENY`?

Reply:
While it doesn't make sense. I think infering that `SHADOW_ALLOW` means action `DENY` is a little counter intuitive no?

Reply:
Would we always require both to be set?

Otherwise

```
default:
  shadowAction: ALLOW
```

would also be valid and mean `action: DENY` so it wouldn't be any less counter intuitive.

Reply:
hmm, but I think it doesn't mean `action: DENY` it means there is no real rule for this set of clients. If you don't have a rule, then yes, the request will be denied, but don't generate rules for that

Reply:
> Is there any reason to set action: ALLOW + shadow: ALLOW or action: DENY + shadow: DENY?

`action: ALLOW + shadow: ALLOW` => want traffic to flow and you don't want shadow engine stats

`action: DENY + shadow: DENY` => want to stop traffic and you want metrics

It all depends on how this translates down to Envoy, if it's a 1:1 mapping it's as I described above.

Reply:
if I understand envoy docs correctly, by default, stats are the same regardless of the `shadow`. But you can add a prefix to shadow stats using [shadow-rules-stat-prefix](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/filters/network/rbac/v3/rbac.proto#envoy-v3-api-field-extensions-filters-network-rbac-v3-rbac-shadow-rules-stat-prefix).

Reply:
> action: ALLOW + shadow: ALLOW => want traffic to flow and you don't want shadow engine stats
> action: DENY + shadow: DENY => want to stop traffic and you want metrics

Do you know why would someone configure this? You get stats from both actions and shadow actions.